### PR TITLE
racy router fix

### PR DESCRIPTION
**Pull request by rogeralsing**
cc @cpx86 and @AsynkronIT/developers.
This fixes the potential racyness in router states.
Does this seem like a fair approach for the other impls too?

### CODE REVIEW

File: proto-router/src/main/kotlin/actor/proto/router/ConsistentHashRouterState.kt [hunk 1]
```diff
@@ -5,22 +5,25 @@ import actor.proto.PID
 import actor.proto.send
 import actor.proto.toShortString
 
+class ConsistentHashRouterRoutees (val hashRing: HashRing,val routeeMap:Map<String, PID>)
+
 internal class ConsistentHashRouterState(private val hash: (String) -> Int, private var replicaCount: Int) : RouterState() {
-    private lateinit var hashRing: HashRing
-    private lateinit var routeeMap: Map<String, PID>
+    private lateinit var routees : ConsistentHashRouterRoutees
```
Comment (rogeralsing) [inline proto-router/src/main/kotlin/actor/proto/router/ConsistentHashRouterState.kt]: Replace the two different fields with one field containing an object with the two previous fields.'
This helps deal with the racyness of reading multiple fields.

File: proto-router/src/main/kotlin/actor/proto/router/ConsistentHashRouterState.kt [hunk 2]
```diff
@@ -5,22 +5,25 @@ import actor.proto.PID
...
+        var nodes = setOf<String>()
         for (pid in routees) {
             val nodeName: String = pid.toShortString()
             nodes += nodeName
             routeeMap += Pair(nodeName, pid)
         }
-        hashRing = HashRing(nodes, hash, replicaCount)
+        val hashRing = HashRing(nodes, hash, replicaCount)
+
+       this.routees  = ConsistentHashRouterRoutees(hashRing,routeeMap)
```
Comment (rogeralsing) [inline proto-router/src/main/kotlin/actor/proto/router/ConsistentHashRouterState.kt]: re-assign the routees field only once everything is recalculated and added to the ConsistentHashRouterRoutees container

File: proto-router/src/main/kotlin/actor/proto/router/ConsistentHashRouterState.kt [hunk 3]
```diff
@@ -30,11 +33,12 @@ internal class ConsistentHashRouterState(private val hash: (String) -> Int, priv
             else -> message
         }
 
+        val r = routees
```
Comment (rogeralsing) [inline proto-router/src/main/kotlin/actor/proto/router/ConsistentHashRouterState.kt]: fetch a reference to the immutable routees object

File: proto-router/src/main/kotlin/actor/proto/router/ConsistentHashRouterState.kt [hunk 4]
```diff
@@ -30,11 +33,12 @@ internal class ConsistentHashRouterState(private val hash: (String) -> Int, priv
...
         when (msg) {
             is Hashable -> {
                 val key = msg.hashBy()
-                val node = hashRing.getNode(key)
-                val routee = routeeMap[node]!!
+                val node = r.hashRing.getNode(key)
```
Comment (rogeralsing) [inline proto-router/src/main/kotlin/actor/proto/router/ConsistentHashRouterState.kt]: Use the same reference for all operations forward.

File: proto-router/src/main/kotlin/actor/proto/router/RandomRouterState.kt [hunk 5]
```diff
@@ -19,8 +19,9 @@ internal class RandomRouterState(seed: Long?) : RouterState() {
     }
 
     override fun routeMessage(message: Any) {
-        val i = random.nextInt(values.count())
-        val pid = values[i]
+        val v = values
```
Comment (rogeralsing) [inline proto-router/src/main/kotlin/actor/proto/router/RandomRouterState.kt]: work with a copy of the routees only

File: proto-router/src/main/kotlin/actor/proto/router/RoundRobinRouterState.kt [hunk 6]
```diff
@@ -18,8 +18,9 @@ internal class RoundRobinRouterState : RouterState() {
     }
 
     override fun routeMessage(message: Any) {
-        val i = currentIndex.getAndIncrement() % values.count()
-        val pid = values[i]
+        val v = values
```
Comment (rogeralsing) [inline proto-router/src/main/kotlin/actor/proto/router/RoundRobinRouterState.kt]: work with a copy of the routees only

**Comment by cpx86** [comment]
I'm thinking it might be be better to do the copy in the router actor instead. That way new (potentially external) routers don't need to know they should copy the routees.

Come to think of it, could it make sense to remove the routee set from the state altogether and instead pass it to the RouteMessage method? That way a router impl just needs to keep track of it's own internal state and know how to apply it's routing algorithm to a set of routees.

**Comment by rogeralsing** [comment]
But the RouteMessage method is called from within the RouterProcess which must get the routees from somewhere.

**Comment by cpx86** [comment]
Yes, but does it have to be the same? OTOH some routers might want their own internal data structure (e.g. consistent hash router) so it's best for the router state to hold the routees actually.

**Comment by cpx86** [comment]
I.e. nevermind that suggestion :)

**Comment by cpx86** [comment]
Anyway, LGTM. Merge at will! :D

Let's also add issues to the other repos to get this fixed.

**Pull request merged**